src/gui/SampleLoader.cpp

```diff
@@ -0,0 +1,138 @@
+/*
+ * SampleLoader.cpp - Static functions that open audio files
+ *
+ * Copyright (c) 2023 saker <sakertooth@gmail.com>
+ *
+ * This file is part of LMMS - https://lmms.io
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public
+ * License as published by the Free Software Foundation; either
+ * version 2 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public
+ * License along with this program (see COPYING); if not, write to the
+ * Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
+ * Boston, MA 02110-1301 USA.
+ *
+ */
+
+#include "SampleLoader.h"
+
+#include <QFileInfo>
+#include <QMessageBox>
+
+#include "ConfigManager.h"
+#include "FileDialog.h"
+#include "GuiApplication.h"
+#include "PathUtil.h"
+#include "SampleDecoder.h"
+#include "Song.h"
+
+namespace lmms::gui {
+QString SampleLoader::openAudioFile(const QString& previousFile)
+{
+	auto openFileDialog = gui::FileDialog(nullptr, QObject::tr("Open audio file"));
+	auto dir = QString{};
+
+	if (!previousFile.isEmpty())
+	{
+		auto file = QString{previousFile};
+		if (QFileInfo{file}.isRelative())
+		{
+			file = ConfigManager::inst()->userSamplesDir() + file;
+			if (!QFileInfo{file}.exists()) { file = ConfigManager::inst()->factorySamplesDir() + previousFile; }
+		}
+
+		dir = QFileInfo{file}.absolutePath();
```

Comment:
Not really. Would probably be more accurate to use `PathUtil::toAbsolute` here too. I notice that similar code was in master, and it seems like I just spruced up the code a bit and didn't notice the simplification that could've been made initially (probably was reluctant to change the code here too to avoid mistakes, my main goal was to just separate it from `SampleBuffer`). Thanks :+1:.